### PR TITLE
optimize WS newhead

### PR DESCRIPTION
**Pull request by codchen**
## Describe your changes and provide context
instead of having each subscription subscribing to tendermint and encode to JSON independently, we will maintain a single subscription with tendermint and a single copy of encoded JSON per head, then fan it out to all outgoing subscriptions

### CODE REVIEW

File: evmrpc/subscribe.go [hunk 1]
```diff
@@ -25,61 +25,98 @@ type SubscriptionAPI struct {
...
-					return
-				}
-				err = notifier.Notify(rpcSub.ID, ethHeader)
-				if err != nil {
-					return
+					break OUTER
 				}
 			case <-rpcSub.Err():
-				return
+				break OUTER
```
Comment (stevenlanders) [inline evmrpc/subscribe.go:101]: is this the same as `continue` here?

File: evmrpc/subscribe.go [hunk 2]
```diff
@@ -25,61 +25,98 @@ type SubscriptionAPI struct {
...
+	logFetcher.filterConfig = filterConfig
+	api := &SubscriptionAPI{
 		tmClient:            tmClient,
 		subscriptionManager: NewSubscriptionManager(tmClient),
 		subscriptonConfig:   subscriptionConfig,
 		logFetcher:          logFetcher,
+		newHeadListenersMtx: &sync.Mutex{},
+		newHeadListeners:    make(map[rpc.ID]chan map[string]interface{}),
 	}
+	id, subCh, err := api.subscriptionManager.Subscribe(context.Background(), NewHeadQueryBuilder(), api.subscriptonConfig.subscriptionCapacity)
```
Comment (stevenlanders) [inline evmrpc/subscribe.go:47]: does this need supervision or some mechanism to retry/reconnect (can the subscription die?)

**Review by stevenlanders — APPROVED** [review]

**Review by philipsu522 — APPROVED** [review]

File: evmrpc/subscribe.go (hunk 1 above)
Comment (codchen) [reply evmrpc/subscribe.go:101]: no we want to break out of the outer loop (instead of the select clause)

File: evmrpc/subscribe.go (hunk 2 above)
Comment (codchen) [reply evmrpc/subscribe.go:47]: I think the subscription manager has a retry mechanism (that's what the `subscription was canceled, resubscribing` log we saw)

**Pull request merged**